Add templating for air-gapped environments

anders-elastisys opened the pull request:
<!-- Choose you PR title carefully as it will be used as the entry in the changelog! -->

> [!warning]
> **This is public repository, ensure not to disclose:**
> - [ ] personal data beyond what is necessary for interacting with this pull request
> - [ ] business confidential information, such as customer names

### What kind of PR is this?

**Required**: Mark one of the following that is applicable:

- [ ] kind/feature       <!-- This PR adds a new feature -->
- [x] kind/improvement   <!-- This PR changes an existing feature -->
- [ ] kind/deprecation   <!-- This PR removes an existing feature -->
- [ ] kind/documentation <!-- This PR contains documentation -->
- [ ] kind/clean-up      <!-- This PR cleans up technical debt -->
- [ ] kind/bug           <!-- This PR fixes a bug -->
- [ ] kind/other         <!-- This PR does something else -->

*Optional*: Mark one or more of the following that are applicable:

> [!important]
> Breaking changes should be marked `kind/admin-change` or `kind/dev-change` depending on type
> Critical security fixes should be marked with `kind/security`

- [x] kind/admin-change  <!-- This PR introduces an admin facing change, add "Platform Administrator notice" section -->
- [ ] kind/dev-change    <!-- This PR introduces a dev facing change, add "Application Developer notice" section -->
- [ ] kind/security      <!-- This PR introduces a critical security fix, add "Security notice" section -->
- [ ] [kind/adr]()       <!-- This PR implements an ADR, add the link -->

<!-- Uncomment the additional sections that applies. -->

<!-- Additional information to be added in the release notes
### Release notes
...
-->

### Platform Administrator notice

There is now an additional `CK8S_FLAVOR` called `air-gapped` with similar settings to `prod` but with some additional configuration to avoid any outgoing traffic to the Internet.

<!-- Add additional information with kind/dev-change
### Application Developer notice
...
-->

<!-- Add additional information with kind/security
### Security notice
...
-->

### What does this PR do / why do we need this PR?

<!-- Add description of the change -->
This PR introduces additional templating to be able to configure CK8s in an air-gapped environment. Most of the config options provides ways of overriding services that normally downloads plugins from the Internet to instead pull from a URL of a server or registry available inside the air-gapped environment.

Also added so that `http_2xx` blackbox monitors work even when `global.verifyTls=false`.

<!-- Add all issues that are fixed by this PR, use "Part of" instead of "Fixes" if you want to keep issues open. -->
- Fixes #

#### Additional information to reviewers

#### Screenshots

#### Checklist

<!-- This section is not added to the changelog or release notes, it is to help you as a contributor and reviewers. -->

- [x] Proper commit message prefix on all commits
  <!-- Example of commit message prefixes:
    - all: changes to multiple areas
    - apps: changes to application running in all clusters
    - apps sc: changes to applications running in the service cluster
    - apps wc: changes to applications running in the workload cluster
    - bin: changes to management binaries or scripts
    - scripts: changes to other scripts
    - config: changes to configuration
    - docs: changes to documentation
    - tests: changes to tests
    - pipeline: changes to the pipeline
    - release: release related
  -->
- Change checks:
  - [x] The change is transparent
  - [ ] The change is disruptive
  - [ ] The change requires no migration steps
  - [ ] The change requires migration steps
- Metrics checks:
  - [ ] The metrics are still exposed and present in Grafana after the change
  - [ ] The metrics names didn't change (Grafana dashboards and Prometheus alerts are not affected)
  - [ ] The metrics names did change (Grafana dashboards and Prometheus alerts were fixed)
- Logs checks:
  - [ ] The logs do not show any errors after the change
- Network Policy checks:
  - [ ] Any changed pod is covered by Network Policies
  - [ ] The change does not cause any dropped packages in the `NetworkPolicy Dashboard`
- Pod Security Policy checks:
  - [ ] Any changed pod is covered by Pod Security Admission
  - [ ] Any changed pod is covered by Gatekeeper Pod Security Policies
  - [ ] The change does not cause any pods to be blocked by Pod Security Admission or Policies
- Falco checks:
  - [ ] The change does not cause any alerts to be generated by Falco
- Audit checks:
  - [ ] The change does not cause any unnecessary Kubernetes audit events
  - [ ] The change requires changes to Kubernetes audit policy
- Bug checks:
  - [ ] The bug fix is covered by regression tests

salehsedghpour commented on helmfile.d/values/prometheus-blackbox-exporter-sc.yaml.gotmpl:
```diff
@@ -4,6 +4,10 @@ pspEnabled: false
+    http_2xx:
+      http:
+        tls_config:
+          insecure_skip_verify: {{ not .Values.global.verifyTls }}
```
Should we (is it possible) add a comment here to discuss that it's not recommended, but might be required in some Air-gapped environments.

anders-elastisys replied:
As this is a global variable that is used to change configurations of other services as well I don't see a need to have a comment for it (not here at least), as this should always be set to true in production environments. I added this so that you don't get alerts when `verifyTls=false`, as in this case you have made a conscious choice to not care about verifying TLS.

salehsedghpour commented on helmfile.d/values/prometheus-blackbox-exporter-wc.yaml.gotmpl:
```diff
@@ -4,6 +4,10 @@ pspEnabled: false
+    http_2xx:
+      http:
+        tls_config:
+          insecure_skip_verify: {{ not .Values.global.verifyTls }}
```
Should we (is it possible) add a comment here to discuss that it's not recommended, but might be required in some Air-gapped environments.

salehsedghpour reviewed (changes requested):
Thanks @anders-elastisys! I was wondering if we should also discuss this feature in Readme files in this repository.

cristiklein reviewed (changes requested):
I love how you give detailed explanation on why each configuration is needed for airgapped.

However, I think this might be better to have as a flavor, alongside alongside `dev` and `prod`. I'm thinking the default values would be unchanged from what we have today, whereas the airgapped flavor would set all the correct values for an airgapped environment.

What do you think?

anders-elastisys commented:
@cristiklein yes I was considering something like that as well, I could look into adding a `offline` or `airgapped` flavor perhaps which atleast sets the override variables to `set-me` so that it is more clear what you need to configure to make it work in an air-gapped environment.

cristiklein commented:
> @cristiklein yes I was considering something like that as well, I could look into adding a `offline` or `airgapped` flavor perhaps which atleast sets the override variables to `set-me` so that it is more clear what you need to configure to make it work in an air-gapped environment.

Thanks @anders-elastisys . Please go with `air-gapped`, since that is what people look for.

anders-elastisys commented:
@salehsedghpour regarding adding a readme in this repo, I am working on public documentation that will be put in the compliantkubernetes repo, I am not sure what the consensus is on having documentation put in this repository. If I add a new flavor as suggested by @cristiklein I can update the main README of this repo to cover using this flavor by setting `CK8S_FLAVOR`.

anders-elastisys commented:
@cristiklein PTAL, I added an `air-gapped` flavor and I also added templating to add custom receivers for alertmanager. Right now the `air-gapped` falvor is the same as `prod` flavor but with additional config options. Maybe there is a better way to do it, such as merging the air-gapped config with prod instead of having the same definitions replicated in two places.

cristiklein reviewed (approved):
Love it, thanks!

cristiklein commented:
@anders-elastisys Just to make sure, have you tried something like `helmfile template` to check that this PR is essentially a no-op for the `dev` and `prod` flavours?

aarnq commented on config/config/common-config.yaml:
```diff
@@ -134,6 +139,13 @@ falco:
       # Can be skipped if eBPF JIT is enabled from outside the container
       hostNetwork: false
 
+    # override the URL used for downloading driver modules, e.g. to use a self hosted file server in an air-gapped environment
+    repoURL: ""
```
Is this only used for the module driver? In that case I'd prefer if it was under the key `module`.

aarnq commented on config/config/common-config.yaml:
```diff
@@ -118,6 +118,11 @@ falco:
       enabled: false
       version: 2.0.0
 
+  artifact:
+    install:
+      # set to false in an air-gapped environment, unless artifacts are self-hosted
+      enabled: true
```
Do we really want this on by default even for regular flavours?
What benefit does it bring compared to the rules bundled with the current image, is there a current version skew?

anders-elastisys replied:
I am not sure if there is any real benefit for having it enabled, I just kept it enabled from previous config. I can maybe try and look into if there is any benefit in keeping it enabled

anders-elastisys replied:
Currently we seem to get falco rules for a specific version [since this PR](https://github.com/elastisys/compliantkubernetes-apps/pull/1843/files) when artifact `install` is enabled (I noticed [v3.0.0](https://github.com/falcosecurity/rules/releases/tag/falco-rules-3.0.0) of falco rules dropped recently).  However, I feel like we can keep it like it is now and maybe look deeper into this in a separate issue, I created [this issue](https://github.com/elastisys/compliantkubernetes-apps/issues/1669) long ago before `artifact.install` was enabled in apps, so maybe it can be investigated as part of that issue instead.

aarnq commented on config/config/sc-config.yaml:
```diff
@@ -718,6 +718,13 @@ opensearch:
   defaultTemplates: true
   additionalTemplates: {}
 
+  plugins:
+    # in an air-gapped environment where the nodes are not connected to the Internet , the following
+    # variable can be used to override where the objectStorage plugin should be downloaded from.
+    objectStorageUrlOverride: # e.g. server.local:8080/repository-s3-2.8.0.zip
+
+    additionalPlugins: []
+
```
```suggestion
  plugins:
    # in an air-gapped environment where the nodes are not connected to the Internet, 
    # this will prevent external/default object storage plugins from being installed
    installExternalObjectStoragePlugin: false
    # or installDefaultObjectStoragePlugin: false
    # or installObjectStoragePlugin: false

    # in an air-gapped environment this can be used to install plugins from known sources
    additionalPlugins: []
      # - server.local:8080/repository-s3-2.8.0.zip
```

Can we rework this a bit and then just reuse the `additionalPlugins`?

aarnq commented on config/config/common-config.yaml:
```diff
@@ -853,6 +865,29 @@ trivy:
   affinity: {}
+
+  # configurations for an offline / air-gapped environment
+  # ref: https://github.com/aquasecurity/trivy/tree/main/helm/trivy#parameters
```
```suggestion
  # ref: https://github.com/aquasecurity/trivy-operator/tree/main/deploy/helm#values
```

aarnq commented on config/config/common-config.yaml:
```diff
@@ -853,6 +865,29 @@ trivy:
+
+  # configurations for an offline / air-gapped environment
+  # ref: https://github.com/aquasecurity/trivy/tree/main/helm/trivy#parameters
+  offline:
+    offlineScan: false
```
```suggestion
  offlineScan:
    enabled: false
```
Or is there other conditions when we would like to edit configuration under `offline` that do not require `offlineScan: true`? Else we should rename the section, and rename the enable flag to `offlineScanEnabled` or similar.

aarnq replied:
Maybe we should call the `offline` key `scanner` instead then, as this mainly applies to the scanner jobs, and the have it as `offlineScanEnabled` to match other config keys.

aarnq commented:
Also, do add the new flavour under `tests/` where applicable. :slightly_smiling_face:

anders-elastisys commented:
@cristiklein  Templating seems to work fine when initializing a new environment using the new `air-gapped` flavor, however, I noticed that the validation done by the scripts in apps did not work as I expected as it only warns if there are `set-me`'s in the override config if they are also set to `set-me` in the default/template config as well. Thus, with how I have configured the variables now you won't get a warning that for example `global.issuer` is not set, which is what I expected. Do we want to keep it this way or do we want to warn for everything that is set to `set-me` for the current ck8s environment config?

aarnq reviewed (approved):
LGTM
